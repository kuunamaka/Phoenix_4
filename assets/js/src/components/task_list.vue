<!-- listing all tasks --> <!-- and feature for deleting the task -->
<template>
  <table>
    <h2>Listing Tasks</h2>
    <li v-for="task in orderedTasks" :key="task.id"> 
      <b>{{ task.name }}</b> (by: <b>{{ task.assignee }}</b>):
      (finished_at: <b>{{ task.finished_at }}</b>,)
      Is_finished?: <b>{{ task.is_finished }}</b>,
      <button v-if="task.is_liked" @click="deleteLike(task)" class="fas fa-heart"></button>
      <button v-else @click="createLike(task)" class="far fa-heart"></button>
      <b>{{ task.favorited_users.length }}</b>,
      <a :href="`/tasks/${ task.id }/edit`" class="button">Edit</a>
    </li>
    <a href="/tasks/new" class="button">Add task</a>
    <a href="/users" class="button">User</a>
  </table>
</template>

<script>
import _ from 'lodash';
import axios from 'axios';
import edit_task from './edit_task'
export default {
  components: { edit_task },
  name: 'task_list',
  data() {
    return {
      tasks: null,
      name: '',
      id: null
    }
  },
  computed: {
    orderedTasks: function() {
      return _.orderBy(this.tasks, 'id')
    }
  },
  methods: {
    async createLike(task){
      try {
        const resp = await axios.post(`/api/tasks` + `/${ task.id }` + `/users`, {
          task: {
            name: this.name,
            id: this.id
          }
        })
        console.log(resp);
        window.location.href = '/tasks';
      }
      catch (e) {
        console.log(e)
      }
    },
    async deleteLike(task){
      try {
        const resp = await axios.delete(`/api/tasks` + `/${ task.id }` + `/users` + `/1`, {
          task: {
            name: this.name,
            id: this.id
          }
        })
        console.log(resp);
        window.location.href = '/tasks';
      }
      catch (e) {
        console.log(e)
      }
    }
  },  
  async mounted() {
    try {
      const resp = await axios.get('/api/tasks')
      this.tasks = resp.data.tasks
    } catch (e) {
      console.log(e)
    }
  }
}
</script>

<style lang="css">
/* like */
.fas {
  background-color: #e0245e;
  border-color: #e0245e;
}
/* unlike */
.far {
  background-color: #bbbbbb;
  border-color: #bbbbbb;
}
</style>
