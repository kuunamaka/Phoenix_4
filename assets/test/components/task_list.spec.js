import { shallowMount } from '@vue/test-utils'
import _ from 'lodash'
import task_list from  '@/src/components/task_list.vue'

const axios = require("axios");
const MockAdapter = require("axios-mock-adapter");
const mock = new  MockAdapter(axios);

mock.onGet("/api/tasks").reply(200, {
  tasks: [
    {id: 1, name: "Test", assignee: "Assignee", is_finished: false, is_liked: false, assignee_id: 2},
    {id: 2, name: "Assignee", assignee: "Test", is_finished: false, is_liked: false, assignee_id: 1}
  ]
});

window = Object.create(window);
const pathname = '/tasks';

Object.defineProperty(window, 'location', {
  value: {
    pathname,
    href: `http://localhost:4001${ pathname }`
  }
});

describe('task_list.vue', () => {
  const wrapper = shallowMount(task_list);
  const vm = wrapper.vm

  it('renders entire component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('order tasks by their id', () => {
    const expectTasks = [
      {id: 1, name: "Test", assignee: "Assignee", is_finished: false, is_liked: false, assignee_id: 2},
      {id: 2, name: "Assignee", assignee: "Test", is_finished: false, is_liked: false, assignee_id: 1}
    ];
    expect(vm.tasks).toEqual(expectTasks)
  })

  describe('createLike/1', () => {
    it('redirect to /tasks', async () => {  
      mock.onPost("/api/tasks/1/users", {task: {id: 1, name: 'Test'}}).reply(200);
      await wrapper.setData({id: 1, name: 'Test'})
        
      await vm.createLike({id: 1, name: 'Test'})

      expect(window.location.href).toEqual('/tasks');
    });
  });

  describe('deleteLike/1', () => {
    it('redirect to /tasks', async () => {  
      mock.onDelete("/api/tasks/1/users/1", {task: {id: 1, name: 'Test'}}).reply(200);
      await wrapper.setData({id: 1, name: 'Test'})
        
      await vm.deleteLike({id: 1, name: 'Test'})

      expect(window.location.href).toEqual('/tasks');
    });
  });
});
