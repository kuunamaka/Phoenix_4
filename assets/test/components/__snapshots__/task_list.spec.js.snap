// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task_list.vue is a Vue instance 1`] = `
<table>
  <h2>
    Listing Tasks
  </h2>
    
  <a
    class="button"
    href="/tasks/new"
  >
    Add task
  </a>
   
  <a
    class="button"
    href="/users"
  >
    User
  </a>
</table>
`;
